### PR TITLE
AMDGPU: Allocate special SGPRs before user SGPR arguments

### LINKED ISSUE #78121: AMDGPU compute functions cannot handle more than 4 SGPR arguments
Any function with an argument list that uses 5 or more SGPRs asserts in implicit argument handling:

```
define void @void_func_v5i32_inreg(<5 x i32> inreg) {
  ret void
}

define void @void_func_5_i32_inreg(i32 inreg, i32 inreg, i32 inreg, i32 inreg, i32 inreg) {
  ret void
}

define void @void_func_a5i32_inreg([5 x i32] inreg) {
  ret void
}

```

Happens for both SelectionDAG and GlobalIsel (with -O0):

```
    frame #4: 0x0000000100ef0634 llc`allocateFixedSGPRInputImpl(CCInfo=0x000000016fdfa678, RC=0x00000001077083f0, Reg=(Reg = 0)) at SIISelLowering.cpp:2229:3
   2226	                                       const TargetRegisterClass *RC,
   2227	                                       MCRegister Reg) {
   2228	  Reg = CCInfo.AllocateReg(Reg);
-> 2229	  assert(Reg != AMDGPU::NoRegister);
   2230	  MachineFunction &MF = CCInfo.getMachineFunction();
   2231	  MF.addLiveIn(Reg, RC);
   2232	}
Target 0: (llc) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #0: 0x000000018fe5911c libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000018fe90cc0 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x000000018fda0a40 libsystem_c.dylib`abort + 180
    frame #3: 0x000000018fd9fd30 libsystem_c.dylib`__assert_rtn + 284
  * frame #4: 0x0000000100ef0634 llc`allocateFixedSGPRInputImpl(CCInfo=0x000000016fdfa678, RC=0x00000001077083f0, Reg=(Reg = 0)) at SIISelLowering.cpp:2229:3
    frame #5: 0x0000000100e9ae4c llc`allocateSGPR64Input(CCInfo=0x000000016fdfa678, Arg=0x000000015505cf9c) at SIISelLowering.cpp:2244:5
    frame #6: 0x0000000100e9ac9c llc`llvm::SITargetLowering::allocateSpecialInputSGPRs(this=0x00000001580185e0, CCInfo=0x000000016fdfa678, MF=0x000000015505bc00, TRI=0x0000000158018330, Info=0x000000015505cef0) const at SIISelLowering.cpp:2296:5
    frame #7: 0x0000000100e9eb0c llc`llvm::SITargetLowering::LowerFormalArguments(this=0x00000001580185e0, Chain=SDValue @ 0x000000016fdfa4b0, CallConv=0, isVarArg=false, Ins=0x000000016fdfb978, D
```L=0x000000016fdfb720, DAG=0x0000000154e084f0, InVals=0x000000016fdfb858) const at SIISelLowering.cpp:2995:7
```

With optimizations enabled, the removed implicit arguments allow it to pass. However this is another bug; they shouldn't be shifting around based on which inputs aren't required

### PR DESCRIPTION
**Pull request by arsenm**
Partially fixes #78121, but there are a few remaining issues:

1. We don't have a path to start packing excess SGPR arguments into VGPRs, and then to the stack.

2. I'm not sure if we should be allowing use of removed implicit arguments as user arguments

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -2990,13 +2994,8 @@ SDValue SITargetLowering::LowerFormalArguments(
   }
 
   // Start adding system SGPRs.
-  if (IsEntryFunc) {
+  if (IsEntryFunc)
     allocateSystemSGPRs(CCInfo, MF, *Info, CallConv, IsGraphics);
-  } else {
-    CCInfo.AllocateReg(Info->getScratchRSrcReg());
-    if (!IsGraphics)
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:2997]: This check limits the call to `allocateSpecialInputSGPRs` from not calling for amdgpu_gfx functions. Are the ABI registers allocated by this function the same for both amdgpu device and gfx functions?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:2997]: No, amdgpu_gfx isn't supposed to get the same inputs. I'm not sure how the workgroup IDs are supposed to be treated for amdgpu_gfx

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:2997]: This is actually a bug fix to better match the DAG behavior

Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:2997]: There is an ongoing patch in the downstream where I enabled architected SGPRs for all callable functions. A patch will be posted in the upstream for gfx12 soon. With that, the call to `allocateSpecialInputSGPRs` needs to be enabled for gfx functions too. We need more clarity on how these special allocations should be treated for gfx functions.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -2782,15 +2782,19 @@ SDValue SITargetLowering::LowerFormalArguments(
...
     // For the fixed ABI, pass workitem IDs in the last argument register.
     allocateSpecialInputVGPRsFixed(CCInfo, MF, *TRI, *Info);
-  }
 
-  if (!IsKernel) {
-    CCAssignFn *AssignFn = CCAssignFnForCall(CallConv, isVarArg);
-    if (!IsGraphics && !Subtarget->enableFlatScratch()) {
+    if (!Subtarget->enableFlatScratch()) {
+      // FIXME: Sink this into allocateSpecialInputSGPRs
       CCInfo.AllocateRegBlock(ArrayRef<MCPhysReg>{AMDGPU::SGPR0, AMDGPU::SGPR1,
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:2788]: Can't we use `CCInfo.AllocateReg(Info->getScratchRSrcReg());` as used in the GlobalISel path?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:2997]: amdgpu_gfx isn't going to have any of the HSA inputs which are passed here. It should only maybe have the workgroup / workitem IDs, but they seem to not

**Review by cdevadas — APPROVED** [review]
LGTM

**Pull request merged**